gulp-tasks/utils/build-browser-package.js

```diff
@@ -0,0 +1,178 @@
+const rename = require('gulp-rename');
+const rollup = require('rollup');
+const rollupStream = require('rollup-stream');
+const source = require('vinyl-source-stream');
+const sourcemaps = require('gulp-sourcemaps');
+
+const constants = require('./constants');
+const logHelper = require('../../infra/utils/log-helper');
+const pkgPathToName = require('./pkg-path-to-name');
+const rollupHelper = require('./rollup-helper');
+
+/*
+ * To test sourcemaps are valid and working, use:
+ * http://paulirish.github.io/source-map-visualization/#custom-choose
+ */
+const ERROR_NO_MODULE_INDEX = `Could not find the modules index.mjs file: `;
+const ERROR_NO_NAMSPACE = oneLine`
+  You must define a 'browserNamespace' parameter in the 'package.json'.
+  Exmaple: 'workbox-precaching' would have a browserNamespace param of
+  'precaching' in 'package.json'. This will be appended to
+  '${constants.NAMESPACE_PREFIX}' meaning developers would use
+  '${constants.NAMESPACE_PREFIX}.precaching' in their
+  JavaScript. Please fix for:
+`;
+
+// This makes Rollup assume workbox-* will be added to the global
+// scope and replace references with the core namespace
+const globals = (moduleId) => {
+  const splitModuleId = moduleId.split('/');
+  if (splitModuleId[0].indexOf('workbox-') !== 0) {
+    throw new Error(`Unknown global module ID: ${moduleId}`);
+  }
+
+  const packageName = splitModuleId.shift();
+  if (splitModuleId.length > 0) {
+    throw new Error(oneLine`
+    All imports of workbox-* modules must be done from the top level export.
+    (i.e. import * from 'workbox-*') This ensures that the browser
+    namespacing works correctly. Please remove '${splitModuleId.join('/')}'
+    from the import '${moduleId}'.
+  `);
+  }
+
+  // Get a package's browserNamespace so we know where it will be
+  // on the global scope (i.e. workbox.<name space>)
+  const packagePath = path.join(__dirname, '..', '..', 'packages', packageName);
+  try {
+    const pkg = require(path.join(packagePath, 'package.json'));
+    return `${constants.NAMESPACE_PREFIX}.${pkg.workbox.browserNamespace}`;
+  } catch (err) {
+    logHelper.error(`Unable to get browserNamespace for package: ` +
+      `'${packageName}'`);
+    logHelper.error(err);
+    throw err;
+  }
+};
+
+// This ensures all workbox-* modules are treated as external and are
+// referenced as globals.
+const externalAndPure = (moduleId) => (moduleId.indexOf('workbox-') === 0);
+
+/*
+ * This function will take an object and generate a friend export
+ * object.
+ * For example, { hello: world, example: { foo: foo } } will output:
+ * "{
+ *   hello: world,
+ *   example: {
+ *     foo,
+ *   },
+ * }"
+ * @param {*} exportObject This is the object that needs to be converted
+ * to a string.
+ * @param {*} levels This value is just used for indenting the code to
+ * ensure the final output is human readable and easy to understand.
+ */
+const convertExportObject = (exportObject, levels = 0) => {
```

Comment:
@gauntface I couldn't tell if/where this was used, but I carried it over from your original build code. Happy to delete it if it's not needed.

Reply:
Please delete - self reference meant it wasn't caught by eslint.